refs #49600 Add iCal view

victor-aston opened the pull request.

victor-aston committed:
refs #49600 - Add iCal view

victor-aston committed:
refs #49600 - Fix route

victor-aston committed:
refs #49600 - Update route

victor-aston committed:
refs #49600 - Add all events

victor-aston committed:
refs #49600 - Add new properties

victor-aston committed:
refs #49600 - Fix description

victor-aston committed:
refs #49600 - Add access check

victor-aston committed:
refs #49600 - Add block

victor-aston committed:
refs #49600 - Add translations

victor-aston committed:
refs #49600 - Fix typo

andrew-sereda commented on composer.json:
```diff
@@ -84,6 +84,7 @@
         "drupal/views_data_export": "^1.3",
         "drupal/views_field_view": "^1.0",
         "drupal/ultimate_cron": "^2.0@alpha",
+        "eluceo/ical": "^0.17",
```
Why we use this version which is more than year old while there is ^2.12.0 from this year?

victor-aston replied:
I've talked with some guys and they suggested using the 0.17 version because not all features are working with a new version right now.
Some contrib modules use 0.17 version as well. e.g. views_ical module.

andrew-sereda commented on modules/living_spaces_event/src/Controller/LivingSpacesEventIcalController.php:
```diff
@@ -0,0 +1,151 @@
+    try {
+      $file = $this->fileRepository->writeData($content, $uri, FileSystemInterface::EXISTS_REPLACE);
+    }
+    catch (\Exception $e) {
```
We should at least add some logs to watchdog about this exception.
Also I don't like the idea users don't see any messaging that something went wrong.
In case of error we should let user know that something went wrong and he should try later, while also logging details to watchdog for us to investigate.

victor-aston replied:
`writeData()` method already adds logs. So, simply added a message.

andrew-sereda commented on modules/living_spaces_event/src/Controller/LivingSpacesEventIcalController.php:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    $filename = "iCal {$group->label()}.ics";
+    $uri = 'public://' . $filename;
```
Would be better to have some dedicated folder for ical exports so they won't overwhelm files root dir

victor-aston replied:
Changed to `public://ical`.

andrew-sereda commented on modules/living_spaces_event/src/Controller/LivingSpacesEventIcalController.php:
```diff
@@ -0,0 +1,151 @@
+    $filepath = $this->fileSystem->realpath('public://') . '/' . $parts[1];
+
+    if (!file_exists($filepath)) {
+      throw new NotFoundHttpException();
```
Messaging for user + logs in watchdog

victor-aston replied:
Added logger and message.

andrew-sereda commented on modules/living_spaces_event/src/Controller/LivingSpacesEventIcalController.php:
```diff
@@ -0,0 +1,151 @@
+  public function access(GroupInterface $group) {
+    $access = $this->currentUser()->hasPermission('administer living spaces event');
+
+    return $access ? AccessResult::allowed() : AccessResult::forbidden();
```
Would be better to have dedicated permission like "access ical export" both global and for groups so we have more flexibility who has access to exports.

victor-aston replied:
Added 'access ical event export' global and group permissions.

andrew-sereda reviewed (changes requested):
Ideally it should have been entity storage query with check access so each user only get events he has access to + batch to generate ical and finally file download.

victor-aston committed:
Merge branch '1.1.x' of github.com:DOM-Digital-Online-Media/living_spaces into ref/49600

victor-aston committed:
refs #49600 - Add logger

victor-aston committed:
refs #49600 - Add access check

victor-aston commented:
> Ideally it should have been entity storage query with check access so each user only get events he has access to + batch to generate ical and finally file download.

Added one more entity query to get a list of event ids to which a user has access and added ids to the filter `$query->condition('id', $event_ids, 'IN');` of the database query. 
In that case, there is no need to load entities and use a batch to get event properties.

victor-aston committed:
refs #49600 - Fix typo
